kerastuner/engine/hypermodel.py

```diff
@@ -32,3 +42,112 @@ class DefaultHyperModel(HyperModel):
     def __init__(self, build, name=None, tunable=True):
         super(DefaultHyperModel, self).__init__(name=name)
         self.build = build
+
+
+class KerasModelWrapper(HyperModel):
```

Comment:
This functionality should be factored as overridable methods on the Tuner, not in a separate wrapper class (that can't be overridden), so as to enable users to override and to reduce indirection.

`build` isn't very likely to overridden, but `compile` is quite likely to be overridden (one example would be making the loss a hyperparameter).

Reply:
IMO conceptually, the functionality provided here (specifying a max model size, allowing failed model builds, using distribution strategies, overriding loss, metrics, etc) is part of building the Model. Having this functionality be part of a HyperModel class buys us the fact that `hypermodel.build` works as expected. Previously, the example below would ignore compiled overrides from the Tuner and would not work with distribution strategies. IMO users should always be using `self.hypermodel.build` to build their Models.

```python
def run_trial(self, trial, x, y):
  model = self.hypermodel.build(trial.hyperparameters)
```

Making loss a HyperParameter is still doable like this:

```python
def my_build_model(hp):
  model = build_model(hp)
  model.compile('adam', loss=hp.Choice('my_loss', ['bce', 'mse']))
  return model
```

A future convenience feature we could add is accepting a `HyperParameter` object for the `Tuner` arguments `loss`, `optimizer`, `metrics`

Reply:
This is especially an issue because the workflow of recreating the best Model and retraining currently requires the use of private methods:

```python
hps = tuner.get_best_hyperparameters()[0]
model = tuner._build_model(hps)
tuner._compile_model(model)
```

Versus in this PR:

```python
hps = tuner.get_best_hyperparameters()
model = tuner.hypermodel.build(hps)
```

Reply:
That makes sense.

Naming note: this class does not wrap a Keras Model, it wraps a HyperModel. It's also a HyperModel subclass itself. Perhaps `KerasHyperModel` would be a better name (conveys that this is a Keras specialization of the generic HyperModel).

Reply:
Done, changed to `KerasHyperModel`